### PR TITLE
New script: Unbound LXC

### PR DESCRIPTION
**Pull request by wimb0**
> [!NOTE]
> We are meticulous when it comes to merging code into the main branch, so please understand that we may reject pull requests that do not meet the project's standards. It's never personal. Also, game-related scripts have a lower chance of being merged.

## Description

This adds Unbound DNS as a standalone LXC.

## Type of change
Please check the relevant option(s):

- [ ] Bug fix (non-breaking change that resolves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (a fix or feature that would cause existing functionality to change unexpectedly)
- [X] New script (a fully functional and thoroughly tested script or set of scripts.)

## Prerequisites
The following efforts must be made for the PR to be considered. Please check when completed:
- [X] Self-review performed (I have reviewed my code, ensuring it follows established patterns and conventions)
- [X] Testing performed (I have tested my changes, ensuring everything works as expected)
- [X] Documentation updated (I have updated any relevant documentation)

## Additional Information (optional)

It is listening at port 5335 for DNS requests.

## Related Pull Requests / Discussions
https://github.com/community-scripts/ProxmoxVE/discussions/530

If there are other pull requests or discussions related to this change, please link them here:
- Related PR #

### CODE REVIEW

File: install/unbound-install.sh [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+wget -qO /var/lib/unbound/root.hints https://www.internic.net/domain/named.root
+chown unbound:unbound /var/lib/unbound/root.hints
+
+touch /var/log/unbound.log
+chown unbound:unbound /var/log/unbound.log
+systemctl enable -q --now unbound
+msg_info "Restarting Unbound to load new config"
+systemctl restart unbound
+msg_ok "Installed Unbound"
```
Comment (MickLesk) [inline install/unbound-install.sh:82]: The service is probably already enabled, or the restart is pointless

File: install/unbound-install.sh [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+msg_ok "Configuring Logrotate"
```
Comment (MickLesk) [inline install/unbound-install.sh:84]: msg_info

File: install/unbound-install.sh [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  sharedscripts
+  create 644
+  postrotate
+    /usr/sbin/unbound-control log_reopen
+  endscript
+}
+EOF
+
+msg_info "Restarting Logrotate"
+systemctl restart logrotate
```
Comment (MickLesk) [inline install/unbound-install.sh:102]: msg_ok below, msg_info remove

File: ct/unbound.sh [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+source <(curl -s https://raw.githubusercontent.com/community-scripts/ProxmoxVE/main/misc/build.func)
+# Copyright (c) 2021-2024 community-scripts ORG
+# Author: wimb0
+# License: MIT
+# https://github.com/community-scripts/ProxmoxVE/raw/main/LICENSE
+
+function header_info {
+clear
+cat <<"EOF"
+     :::    ::: ::::    ::: :::::::::   ::::::::  :::    ::: ::::    ::: :::::::::
```
Comment (MickLesk) [inline ct/unbound.sh:11]: Slant Font.

https://www.patorjk.com/software/taag/#p=display&f=Slant&t=Unbound

**Review by MickLesk — CHANGES_REQUESTED** [review]
I have left a few comments. Revise the script first, then I'll take another look. :smile:

As a Big Note: 
The .json for the web page is missing. You can also generate this on our website:
https://community-scripts.github.io/ProxmoxVE/json-editor

File: install/unbound-install.sh [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  logfile: /var/log/unbound.log
+  verbosity: 1
+  log-queries: yes
+  statistics-interval: 0
+  extended-statistics: yes
+  harden-below-nxdomain: yes
+EOF
+
+wget -qO /var/lib/unbound/root.hints https://www.internic.net/domain/named.root
+chown unbound:unbound /var/lib/unbound/root.hints
```
[the remark this answers is not in the document]
Comment (wimb0) [reply install/unbound-install.sh:75]: 1. Update Root hints from Internic (This file holds the information on root name servers needed to initialize cache of Internet domain name servers)
2. This user is created by the unbound apt package

File: install/unbound-install.sh [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+touch /var/log/unbound.log
+chown unbound:unbound /var/log/unbound.log
```
[the remark this answers is not in the document]
Comment (wimb0) [reply install/unbound-install.sh:78]: The unbound process runs under the unbound user. if we don't touch the file and change ownership, unbound cannot write logfiles.

File: install/unbound-install.sh [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+source /dev/stdin <<< "$FUNCTIONS_FILE_PATH"
+color
+verb_ip6
+catch_errors
+setting_up_container
+network_check
+update_os
+
+msg_info "Installing Unbound"
```
[the remark this answers is not in the document]
Comment (wimb0) [reply install/unbound-install.sh:16]: added dependencies

File: ct/unbound.sh (hunk 4 above)
Comment (wimb0) [reply ct/unbound.sh]: fixed 😊

**Comment by wimb0** [comment]
Fix requested changes.

File: install/unbound-install.sh [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  logfile: /var/log/unbound.log
+  verbosity: 1
+  log-queries: yes
+  statistics-interval: 0
+  extended-statistics: yes
+  harden-below-nxdomain: yes
+EOF
+
+# Update Root hints from Internic (This file holds the information on root name servers needed to initialize cache of Internet domain name servers)
+wget -qO /var/lib/unbound/root.hints https://www.internic.net/domain/named.root
```
Comment (havardthom) [inline install/unbound-install.sh:86]:
```suggestion
wget -qO /var/lib/unbound/root.hints https://www.internic.net/domain/named.root
```

image

File: install/unbound-install.sh [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+# Set unbound user as owner of the root hints file
+chown unbound:unbound /var/lib/unbound/root.hints
+
+touch /var/log/unbound.log
+chown unbound:unbound /var/log/unbound.log
+
+systemctl restart unbound
+msg_ok "Installed Unbound"
+
+msg_ok "Configuring Logrotate"
```
Comment (havardthom) [inline install/unbound-install.sh:96]:
```suggestion
msg_info "Configuring Logrotate"
```

File: install/unbound-install.sh [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+  sharedscripts
+  create 644
+  postrotate
+    /usr/sbin/unbound-control log_reopen
+  endscript
+}
+EOF
+
+systemctl restart logrotate
+msg_ok "Restarted Logrotate"
```
Comment (havardthom) [inline install/unbound-install.sh:114]:
```suggestion
msg_ok "Configured Logrotate"
```

File: install/unbound-install.sh [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+  mc 
+msg_ok "Installed Dependencies"
+
+msg_info "Installing Unbound"
+$STD apt-get install -y \
+  unbound \
+  unbound-host
+msg_info "Installed Unbound"
+
+cat <<EOF >/etc/unbound/unbound.conf.d/unbound-lxe.conf
```
Comment (havardthom) [inline install/unbound-install.sh:29]:
```suggestion
cat <<EOF >/etc/unbound/unbound.conf.d/unbound.conf
```

File: install/unbound-install.sh [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+server:
+  verbosity: 0
```
Comment (havardthom) [inline install/unbound-install.sh:31]:
```suggestion
  verbosity: 1
```
What is this configuration file based on? It should be a minimal working configuration with mostly defaults

File: install/unbound-install.sh [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+  private-address: 10.0.0.0/8
+  private-address: fd00::/8
+  private-address: fe80::/10
+  access-control: 192.168.0.0/16 allow
+  access-control: 172.16.0.0/12 allow
+  access-control: 10.0.0.0/8 allow
+  access-control: 127.0.0.1/32 allow
+  chroot: ""
+  logfile: /var/log/unbound.log
+  verbosity: 1
```
Comment (havardthom) [inline install/unbound-install.sh:78]:
```suggestion
```

**Review by havardthom — DISMISSED** [review]

File: install/unbound-install.sh (hunk 8 above)
Comment (wimb0) [reply install/unbound-install.sh]: Some Unbound install guides advise updating the root.hints files. Pi-hole lists it as optional.
image

I've checked on a fresh Unbound install via apt without updating the files manually, and the root.hints file is up-to-date.
I removed the download and config of the root.hints file.

File: install/unbound-install.sh (hunk 13 above)
Comment (wimb0) [reply install/unbound-install.sh]: Fixed. Sorry, that was a leftover from testing.

File: install/unbound-install.sh (hunk 12 above)
Comment (wimb0) [reply install/unbound-install.sh:31]: I based this config on the configuration from Pi-hole (https://docs.pi-hole.net/guides/dns/unbound/) and used parts of https://github.com/saint-lascivious/unbound-config

File: install/unbound-install.sh [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+  private-address: 10.0.0.0/8
+  private-address: fd00::/8
+  private-address: fe80::/10
+  access-control: 192.168.0.0/16 allow
+  access-control: 172.16.0.0/12 allow
+  access-control: 10.0.0.0/8 allow
+  access-control: 127.0.0.1/32 allow
+  chroot: ""
+  logfile: /var/log/unbound.log
+  log-queries: yes
```
Comment (havardthom) [inline install/unbound-install.sh:77]: Is this needed when verbosity is 0?

File: install/unbound-install.sh [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+  statistics-interval: 0
```
Comment (havardthom) [inline install/unbound-install.sh:78]:
```suggestion
```
default 0

File: install/unbound-install.sh [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+  extended-statistics: yes
```
Comment (havardthom) [inline install/unbound-install.sh:79]: is this needed when `statistics-interval` is 0?

File: install/unbound-install.sh [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+  harden-below-nxdomain: yes
```
Comment (havardthom) [inline install/unbound-install.sh:80]:
```suggestion
```
default yes

File: json/unbound.json [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+            }
+        }
+    ],
+    "default_credentials": {
+        "username": null,
+        "password": null
+    },
+    "notes": [
+        {
+            "text": "Unbound listens to DNS requests at port 5335",
```
Comment (havardthom) [inline json/unbound.json:35]: We can remove this note. port is already shown through `interface_port`

File: json/unbound.json [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+            "type": "warning"
+        },
+        {
+            "text": "Unbound Configuration Path: `/etc/unbound/unbound.conf.d/unbound-lxe.conf`",
```
Comment (havardthom) [inline json/unbound.json:39]:
```suggestion
            "text": "Unbound Configuration Path: `/etc/unbound/unbound.conf.d/unbound.conf`",
```

**Review by havardthom — DISMISSED** [review]

File: install/unbound-install.sh [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+  do-ip6: no
+  do-ip4: yes
+  do-udp: yes
+  do-tcp: yes
+  num-threads: 1
+  hide-identity: yes
+  hide-version: yes
+  harden-glue: yes
+  harden-dnssec-stripped: yes
+  harden-referral-path: yes
```
[the remark this answers is not in the document]
Comment (wimb0) [reply install/unbound-install.sh:42]: I've added this setting to harden Unbound and enforce DNSSEC.

File: json/unbound.json (hunk 18 above)
Comment (wimb0) [reply json/unbound.json]: Fixed.

File: install/unbound-install.sh [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+  rrset-cache-size: 256m
+  msg-cache-size: 128m
+  so-rcvbuf: 1m
+  private-address: 192.168.0.0/16
+  private-address: 169.254.0.0/16
+  private-address: 172.16.0.0/12
+  private-address: 10.0.0.0/8
+  private-address: fd00::/8
+  private-address: fe80::/10
+  access-control: 192.168.0.0/16 allow
```
[the remark this answers is not in the document]
Comment (wimb0) [reply install/unbound-install.sh:62]: `access-control:` control which clients are allowed to make (recursive) queries to this server.
I've added local IP ranges (rfc1918), so clients can use Unbound as DNS server.
By default everything is refused, except for localhost, if we don't configure `access-control:`

File: install/unbound-install.sh [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+  harden-dnssec-stripped: yes
+  harden-referral-path: yes
+  use-caps-for-id: no
+  harden-algo-downgrade: no
+  qname-minimisation: yes
+  aggressive-nsec: yes
+  rrset-roundrobin: yes
+  cache-min-ttl: 300
+  cache-max-ttl: 14400
+  msg-cache-slabs: 8
```
[the remark this answers is not in the document]
Comment (havardthom) [reply install/unbound-install.sh]: Remove other cache slabs overwrites and keep default

File: install/unbound-install.sh (hunk 14 above)
Comment (wimb0) [reply install/unbound-install.sh]: Yes, Unboud still logs errors.
We could use syslog, which is default, but I think a separate logfile is nicer.

File: install/unbound-install.sh (hunk 16 above)
Comment (wimb0) [reply install/unbound-install.sh]: Removed `statistics-interval` and `extended-statistics`

File: install/unbound-install.sh (hunk 14 above)
Comment (havardthom) [reply install/unbound-install.sh]: You can add `log-queries: yes` back then

Comment (wimb0) [reply install/unbound-install.sh]:
```
# print one line with time, IP, name, type, class for every query.
# log-queries: no
```

Default is no, not sure if it is a good idea to log every query by default.

**Review by havardthom — APPROVED** [review]

**Review by MickLesk — APPROVED** [review]

**Pull request merged**